src/utils/files/glob-source.js

```diff
@@ -0,0 +1,111 @@
+const cat = require('pull-cat')
+const defer = require('pull-defer')
+const pushable = require('pull-pushable')
+const map = require('async/map')
+const parallel = require('async/parallel')
+const errCode = require('err-code')
+
+/**
+* Create a pull stream source that can be piped to ipfs.addPullStream for the
+* provided file paths.
+*
+* @param ...paths {String} File system path(s) to glob from
+* @param [options] {Object} Optional options
+* @param [options.recursive] {Boolean} Recursively glob all paths in directories
+* @param [options.hidden] {Boolean} Include .dot files in matched paths
+* @param [options.ignore] {Array<String>} Glob paths to ignore
+* @param [options.followSymlinks] {Boolean} follow symlinks
+*/
+module.exports = (...args) => {
+  const options = isString(args[args.length - 1]) ? {} : args.pop()
+  const paths = args
+  const deferred = defer.source()
+
+  const globSourceOptions = {
+    recursive: options.recursive,
+    glob: {
+      dot: Boolean(options.hidden),
+      ignore: Array.isArray(options.ignore) ? options.ignore : [],
+      follow: options.followSymlinks != null ? options.followSymlinks : true
+    }
+  }
+
+  // Check the input paths comply with options.recursive and convert to glob sources
+  map(paths, normalizePathWithType, (err, results) => {
+    if (err) return deferred.abort(err)
+
+    try {
+      const sources = results.map(res => toGlobSource(res, globSourceOptions))
+      return deferred.resolve(cat(sources))
+    } catch (err) {
+      return deferred.abort(err)
+    }
+  })
+
+  return pull(
+    deferred,
+    pull.map(({ path, contentPath }) => ({
+      path,
+      content: fs.createReadStream(contentPath)
+    }))
+  )
+}
+
+function toGlobSource ({ path, type }, options) {
+  options = options || {}
+
+  const baseName = Path.basename(path)
+
+  if (type === 'file') {
+    return pull.values([{ path: baseName, contentPath: path }])
+  }
+
+  if (type === 'dir' && !options.recursive) {
+    throw errCode(
+      new Error(`'${path}' is a directory and recursive option not set`),
+      'ERR_DIR_NON_RECURSIVE',
+      { path }
+    )
+  }
+
+  const globOptions = Object.assign({}, options.glob, {
+    cwd: path,
+    nodir: true,
+    realpath: false,
+    absolute: false
+  })
+
+  // TODO: want to use pull-glob but it doesn't have the features...
+  const pusher = pushable()
+
+  glob('**/*', globOptions)
+    .on('match', m => pusher.push(m))
+    .on('end', () => pusher.end())
+    .on('abort', () => pusher.end())
+    .on('error', err => pusher.end(err))
+
+  return pull(
+    pusher,
+    pull.map(p => ({
+      path: Path.join(baseName, p),
+      contentPath: Path.join(path, p)
+    }))
+  )
+}
+
+function normalizePathWithType (path, cb) {
+  parallel({
+    stat: cb => fs.stat(path, cb),
+    realpath: cb => fs.realpath(path, cb)
+  }, (err, res) => {
+    if (err) return cb(err)
+    cb(null, { path: res.realpath, type: res.stat.isDirectory() ? 'dir' : 'file' })
```

Comment:
What happens if `path` is a socket, symbolic link, block device, etc?

Reply:
Undefined. Needs tests! :D

Reply:
`fs.realpath` has been removed - it wasn't necessary.

Symlinks work fine for both files and directories. I've added some tests here: https://github.com/ipfs/interface-ipfs-core/pull/421 but I'm not sure if these will work on windows.

For a socket or block device, it essentially boils down to if `fs.createReadStream` can be used to read it then it can be added https://github.com/ipfs/js-ipfs/blob/4258046dadc357e87dd3ba5db445c269dde09c33/src/utils/files/glob-source.js#L55

I'm unsure how I'd test these edge cases right now...but it's worth noting the behaviour would be the [same as before](https://github.com/ipfs/js-ipfs/blob/6d46e2e9bf5d36dc7f8dcbc92b97347e193719d3/src/cli/commands/add.js#L79) (in the case of `ipfs add`) and there's currently no special provisions for these things in [`addFromFs` in `js-ipfs-http-client`](https://github.com/ipfs/js-ipfs-http-client/blob/master/src/utils/prepare-file.js).

Given that all the current tests are passing and CI is green, I'm feeling as though we should get this merged so that we can release 0.34 and deal with bug reports for sockets or block devices when they come in. Does that sound sensible?